File: runtime/CSharp/Antlr4.Runtime/Atn/PrecedencePredicateTransition.cs
// Copyright (c) Terence Parr, Sam Harwell. All Rights Reserved.
// Licensed under the BSD License. See LICENSE.txt in the project root for license information.

using Antlr4.Runtime.Sharpen;

namespace Antlr4.Runtime.Atn
{
    /// <author>Sam Harwell</author>
    public sealed class PrecedencePredicateTransition : AbstractPredicateTransition
    {
        public readonly int precedence;

        public PrecedencePredicateTransition(ATNState target, int precedence)
            : base(target)
        {
            this.precedence = precedence;
        }

        public override Antlr4.Runtime.Atn.TransitionType TransitionType
        {
            get
            {
                return Antlr4.Runtime.Atn.TransitionType.Precedence;
            }
        }

        public override bool IsEpsilon
        {
            get
            {
                return true;
            }
        }

        public override bool Matches(int symbol, int minVocabSymbol, int maxVocabSymbol)
        {
            return false;
        }

        public SemanticContext.PrecedencePredicate Predicate
        {
            get
            {
                return new SemanticContext.PrecedencePredicate(precedence);
            }
        }

        public override string ToString()
        {
            return precedence + " >= _p";
        }
    }
}

File: runtime/Java/src/org/antlr/v4/runtime/atn/PrecedencePredicateTransition.java
/*
 * Copyright (c) 2012 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD-3-Clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

package org.antlr.v4.runtime.atn;

import org.antlr.v4.runtime.misc.NotNull;

/**
 *
 * @author Sam Harwell
 */
public final class PrecedencePredicateTransition extends AbstractPredicateTransition {
	public final int precedence;

	public PrecedencePredicateTransition(@NotNull ATNState target, int precedence) {
		super(target);
		this.precedence = precedence;
	}

	@Override
	public int getSerializationType() {
		return PRECEDENCE;
	}

	@Override
	public boolean isEpsilon() {
		return true;
	}

	@Override
	public boolean matches(int symbol, int minVocabSymbol, int maxVocabSymbol) {
		return false;
	}

	public SemanticContext.PrecedencePredicate getPredicate() {
		return new SemanticContext.PrecedencePredicate(precedence);
	}

	@Override
	public String toString() {
		return precedence + " >= _p";
	}

}
